# Swetha Chandrasekar
# 012497628
# Exploring the use of word vectors and doc vectors as features for predicting fake news
# <u>Business problem </u>


<h1> Given a 'text' of statement made by a politician and 'the party affliation', predict whether the news is fake news or not </h1>
<h2> dataset: LIAR dataset </h2>
<h3>labelling <br>

<font color = red>0 - news labelled as [ false, barely true ] </font> <b> in LIAR dataset </b> <br>
<font color=green> 1 - new labelled as [true, mostly true, half true] <b> in LIAR dataset </b></font>
</h3>

In [47]:
lines = open('train.tsv','r',encoding='utf-8').readlines()

In [48]:
#Size of the dataset LIAR
len(lines)


10269

In [49]:
d =lines[0].strip().split('\t')
for i in range(0,len(d)):
    print (i,d[i])

0 2635.json
1 false
2 Says the Annies List political group supports third-trimester abortions on demand.
3 abortion
4 dwayne-bohac
5 State representative
6 Texas
7 republican
8 0
9 1
10 0
11 0
12 0
13 a mailer


<h1> Label 1 for true , Label 0 for fake news </h1>
<h2> true and mostly-true news are marked as 1 and false, barely-true and half-true are marked as 0 </h2>

In [50]:
true_data = []
false_data = []
overall_data = []
label = []
for i in range(0,len(lines)):
    line = lines[i].strip().lower().split()
    overall_data.append(lines[i].strip().lower())
    if (line[1]== 'true' or line[1] =='mostly-true'):
        true_data.append(lines[i].strip().lower())
        label.append(1)
    else:
        false_data.append(lines[i].strip().lower())
        label.append(0)
## Please note overall data still contains all of training data

# Data preprocessing
1. word tokenization
2. stop words removal
3. punctuation removal


<h1>use nltk word tokenize and remove stop words </h1>



In [51]:
##use nltk word tokenize and remove stop words
import nltk;
from nltk import word_tokenize
from nltk.corpus import stopwords
import string
stop = stopwords.words('english') + list(string.punctuation)
data = []
for line in overall_data:
    text = line.split('\t')[2].lower()
    party = line.split('\t')[7].lower()
    line = party + ' ' + text 
    tokens = [i for i in word_tokenize(line.strip().lower()) if i not in stop]
    bigrams = []
    for i in range(0,len(tokens)-1):
        bigram = tokens[i]+'-'+tokens[i+1]
        bigrams.append(bigram)
    for bigram in bigrams:
        tokens.append(bigram)
        
    data.append(tokens)
data[0]

['republican',
 'says',
 'annies',
 'list',
 'political',
 'group',
 'supports',
 'third-trimester',
 'abortions',
 'demand',
 'republican-says',
 'says-annies',
 'annies-list',
 'list-political',
 'political-group',
 'group-supports',
 'supports-third-trimester',
 'third-trimester-abortions',
 'abortions-demand']

<h1> Use Gensim to create word vectors from the dataset </h1>

In [52]:
# code referenced from https://machinelearningmastery.com/develop-word-embeddings-python-gensim/
from gensim.models import Word2Vec
model = Word2Vec(data, min_count=1)
# summarize the loaded model
print(model)
# summarize vocabulary
words = list(model.wv.vocab)
#print(words)
# access vector for one word
print(model['democrat'])
# save model
model.save('overall_model.bin')
# load model
overall_model = Word2Vec.load('overall_model.bin')
print(overall_model)
word_vector = overall_model

Word2Vec(vocab=92977, size=100, alpha=0.025)
[ 0.3042278  -0.19084701  0.30704698 -0.04614774  0.30127057 -0.39753827
  0.21099289  0.04760346 -0.09727817  0.00110407 -0.1509957  -0.58583367
  0.19755447  0.15945695  0.02719315 -0.05512587 -0.4164396   0.7089397
 -0.10720449  0.2343237   0.10276304  0.29957622  0.44162408 -0.05149297
  0.07815396 -0.3893264   0.10842299 -0.2528649  -0.55113417  0.21559718
 -0.35969508 -0.0623625  -0.1495717  -0.43451336  0.09172938  0.15576635
  0.03578302  0.5117066  -0.9358748   0.01989187 -0.0129511  -0.09246799
  0.04944352  0.16114403  0.34654868 -0.22775806 -0.01824609 -0.21470852
  0.41621545  0.295388   -0.17595226  0.01865412  0.4773739  -0.21952885
  0.00775523 -0.00489549 -0.03777182 -0.07143733 -0.0597306   0.23876062
 -0.05098007  0.30073833 -0.40168673  0.08237503  0.17868659 -0.12636824
 -0.25743777 -0.21753372 -0.16529737  0.4817043   0.5583272   0.00732034
  0.00345164 -0.05041166 -0.20952019  0.33536816 -0.25522247  0.54077387
  0.300

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


Word2Vec(vocab=92977, size=100, alpha=0.025)


In [53]:
word_vector["obama"]

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


array([ 0.2853234 , -0.48534012,  0.5000886 , -0.2689489 ,  0.7408342 ,
       -0.5546253 ,  0.3838437 , -0.14750402,  0.15514989,  0.0542842 ,
        0.47308603, -0.7753996 ,  0.22022419,  0.392062  ,  0.44076765,
       -0.3084322 , -0.37180814,  1.0051496 , -0.4380037 , -0.544555  ,
        0.53491664,  0.5341076 ,  0.8476037 , -0.16393945,  0.3179823 ,
       -0.6324809 ,  0.82084453,  0.09179328, -1.0027754 ,  0.15624635,
       -0.14643313,  0.25749782, -0.3352869 , -0.89560634,  0.18808894,
        0.32283223, -0.53243345,  0.91719013, -1.1129243 , -0.15304033,
        0.64297837,  0.11762323, -0.35534397,  0.38519317,  0.36889327,
       -0.48139963, -0.5959327 , -0.5698601 ,  0.03823274,  0.46288544,
       -0.18475063, -0.08369005,  0.46727866, -0.2556913 , -0.04049212,
        0.28839445, -0.20834452,  0.02781974, -0.40480906, -0.00590318,
        0.21219815, -0.017327  , -0.3677732 , -0.47276872,  0.3141341 ,
        0.47274998, -0.3129866 , -0.519324  , -0.81234276,  0.51

## Word vectors represent latent signals projecting word as a vector in an n-dimensional space
## Let us use word vectors directly as features to predict if the news is fake or not
word vector simply gives array of numbers of size n. <br> <b>
<font color=green>We can directly use the sum of word vectors as features.</font> <br> </b>
## 1. sentences are made of multiple words, 
## 2. then we simply add the vectors of each word and get resultant vector for that sentence and use that as feature
## example vector('obama is socialist') = vector['obama'] + vector['is']+  vector['socialist']


<h1> <font color=red> This may not be the right approach as doc2vec is a better approach. This was suggested by professor in class to compare it with doc2vec. </font> </h1>

In [54]:
X = []
for line in data:
    vectors = []
    for word in line:
        if word in overall_model:
            vector = overall_model[word]
            vectors.append(vector)
    word_vector = []

    for i in range(0,len(vectors[0])):
        val = 0.0
        for j in range(0,len(vectors)):
            val += vectors[j][i]
        word_vector.append(val)
    X.append(word_vector)
        


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:5: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  """
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:6: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  


# Run Logistic regression on features from word vector to predict label (0,1)
# 0 is fake news, 1 is truthful news

In [65]:
import pandas as pd
X_df = pd.DataFrame(X)
Y = pd.DataFrame(label)
scaled_X = X_df
scaled_Y = Y
from sklearn.model_selection import train_test_split
from sklearn import linear_model
X_train, X_test, y_train, y_test = train_test_split(scaled_X, scaled_Y, test_size=0.3, random_state=0)
lm = linear_model.LogisticRegression(verbose=1)
model = lm.fit(X_train, y_train)
print (model)
predictions = lm.predict(X_test)

print ("Score:", model.score(X_test, y_test))
w2vecmodel = model

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


[LibLinear]LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=1, warm_start=False)
Score: 0.6416747809152873


## Logistic regression was trained on word vectors and it has accuracy of 0.64 in predicting the fake news on test set. let us test on a sample sentence 
# lets say text= "republican says obama supports third trimester abortion"

In [56]:
testing= "republican says obama supports third trimester abortion"

import numpy as np
def getVector(s):
    s = s.lower()
    words = word_tokenize(s)
    vectors = []
    for word in words:
        if word in overall_model:
            vectors.append(overall_model[word])
    sum_vector = []
    for i in range(0,len(vectors[0])):
        val = 0.0
        for j in range(0,len(vectors)):
            val += vectors[j][i]
        sum_vector.append(val)
    return pd.DataFrame([sum_vector])

testX = getVector(testing)
result_score = model.predict(testX)
model.predict_proba(testX)
### 0.71 probability that news belong to class 0 which is fake news

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:9: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if __name__ == '__main__':
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


array([[0.75521006, 0.24478994]])

## Class 1 probability is 0.24 and class 0 probability is 0.75 , 


<h1>so its likely fake</h1>


<h2><font color=red> fake news: </font>  <font color=red> "republican says obama supports third trimester abortion" </font> </h2>

# So far, we tried word2vec and summed the word vectors of individual words to get vector for each text

# Let's try doc2vec to see if embeddings for text as a whole is performs better than word2vec

In [57]:
#https://medium.com/@mishra.thedeepak/doc2vec-simple-implementation-example-df2afbbfbad5
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.tokenize import word_tokenize
data = []
for line in overall_data:
    T = line.strip().split('\t')
    text = T[2]
    party = T[7]
    text = party + ' ' + text
    data.append(text.lower())
tagged_data = [TaggedDocument(words=word_tokenize(_d.lower()), tags=[str(i)]) for i, _d in enumerate(data)]
data[0]

'republican says the annies list political group supports third-trimester abortions on demand.'

In [58]:
max_epochs = 10
vec_size = 20
alpha = 0.025

model = Doc2Vec(size=vec_size,
                alpha=alpha, 
                min_alpha=0.00025,
                min_count=1,
                dm =1)
  
model.build_vocab(tagged_data)

for epoch in range(max_epochs):
    print('iteration {0}'.format(epoch))
    model.train(tagged_data,
                total_examples=model.corpus_count,
                epochs=model.iter)
    # decrease the learning rate
    model.alpha -= 0.0002
    # fix the learning rate, no decay
    model.min_alpha = model.alpha

model.save("d2v.model")
print("Model Saved")

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/gensim/models/doc2vec.py:570: UserWarning: The parameter `size` is deprecated, will be removed in 4.0.0, use `vector_size` instead.
  warnings.warn("The parameter `size` is deprecated, will be removed in 4.0.0, use `vector_size` instead.")


iteration 0


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:17: DeprecationWarning: Call to deprecated `iter` (Attribute will be removed in 4.0.0, use self.epochs instead).


iteration 1
iteration 2
iteration 3
iteration 4
iteration 5
iteration 6
iteration 7
iteration 8
iteration 9
Model Saved


In [59]:
from gensim.models.doc2vec import Doc2Vec

model= Doc2Vec.load("d2v.model")
#to find the vector of a document which is not in training data
test_data = word_tokenize("republicans say obama supports third trimester abortion".lower())
v1 = model.infer_vector(test_data)
v1


array([ 0.00403764, -0.07682121, -0.01614466, -0.0349686 , -0.04510788,
       -0.06150846,  0.18929797,  0.01581638,  0.02578834,  0.19873458,
        0.20016974, -0.24335104,  0.11643776,  0.09796631, -0.13430543,
        0.13469905,  0.01366164, -0.07000271,  0.06492338,  0.19003768],
      dtype=float32)

In [60]:
doc2vecFeatures = []
for line in overall_data:
    T = line.strip().split('\t')
    text = T[2]
    party = T[7]
    text = party + ' ' + text
    doc2vecFeatures.append(model.infer_vector(word_tokenize(text)))

# Let us use doc2vec vector embeddings directly as a feature and measure its performance against word2vec vector sums

In [61]:
import pandas as pd
X_df = pd.DataFrame(doc2vecFeatures)
Y = pd.DataFrame(label)
scaled_X = X_df
scaled_Y = Y
from sklearn.model_selection import train_test_split
from sklearn import linear_model
X_train, X_test, y_train, y_test = train_test_split(scaled_X, scaled_Y, test_size=0.3, random_state=0)
lm = linear_model.LogisticRegression(verbose=1)
model = lm.fit(X_train, y_train)
print (model)
predictions = lm.predict(X_test)

print ("Score:", model.score(X_test, y_test))

[LibLinear]LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=1, warm_start=False)
Score: 0.6410256410256411


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [62]:
text = "republican says obama supports third trimester abortion"
dvmodel= Doc2Vec.load("d2v.model")
test_sentence= [dvmodel.infer_vector(word_tokenize(text))]

In [63]:
predicted = model.predict(pd.DataFrame(test_sentence))
predicted_probabilities = model.predict_proba(pd.DataFrame(test_sentence))

In [64]:
print ("Doc2Vec prediction")
print ("=====================")
print ("Predicted class:",predicted, "Predicted probabilities for class[0,1]:",predicted_probabilities)
print ("Length of embeddings from Doc2Vec:", len(dvmodel.infer_vector(word_tokenize(text))))
df = getVector(text)
print ("Length of embeddings from Word2Vec embeddings:",len(df.columns))
print ("====================================================================")


Doc2Vec prediction
Predicted class: [0] Predicted probabilities for class[0,1]: [[0.63726274 0.36273726]]
Length of embeddings from Doc2Vec: 20
Length of embeddings from Word2Vec embeddings: 100


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:9: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if __name__ == '__main__':
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


## Class 1 probability is 0.36 and class 0 probability is 0.63 , 


<h1>so its likely fake </h1> 


<h1><font color=red> fake news </font> : <font color=red> "republican says obama supports third trimester abortion" </font> </h1>

# Accuracy in predicting fake news based on 
## Doc2Vec embeddings: 0.6410
##  Word2Vec embeddings: 0.6416

# word2vec embeddings with vector summation seems to perform similar to Doc2Vec embeddings
# it could be because word2vec is generating vectors of size <u>100</u> and while doc2vec is generating vectors of size <u>20.</u> 
# Doc 2 vec seem to get the same accuracy with 20 embeddings that word2vec does with 100 embeddings.